Enable incremental Pan/Tilt/Zoom capability to Amcrest/Dahua cameras

GaryOkie opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add new "**amcrest.ptz_control**" service to specify a PTZ camera movement or zoom direction (up, down, right, left, right_up, right_down, left_up, left_down, zoom_in, zoom_out).   An optional _travel_time_ attribute specifies the amount of movement between start/stop.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] **New feature** (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example snippet from lovelace picture_elements card
# (full example will be provided in documentation)
```
    tap_action:
      action: call-service
      service: amcrest.ptz_control
      service_data:
        entity_id: camera.house_front
        movement: up
        travel_time: .4

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #   N/A
- This PR is related to issue:   [Amcrest IP Camera Enhancements - PTZ Control](https://community.home-assistant.io/t/amcrest-ip-camera-component-enhancements-ptz-control-and-audio-streaming/76655/262)
Note: This link also shows this enhancement with PTZ arrow control overlay

- Link to documentation pull request: 
**submitted PR  [12383](https://github.com/home-assistant/home-assistant.io/pull/12383#)** 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

GaryOkie committed:
Update Amcrest services.yaml to add PTZ camera movement/zoom control

GaryOkie committed:
Update Amcrest camera.py to enable incremental PTZ camera move/zoom control

Add ability to pan/tilt PT camera right/left/up/down and to zoom in/out a PTZ camera in travel time increments.

GaryOkie committed:
Update camera.py

GaryOkie committed:
Update camera.py to fix lint "style" errors

GaryOkie committed:
Update camera.py to fix flake8 errors

GaryOkie committed:
Update camera.py - fix continuation/indentation flake warning

GaryOkie commented:
I believe I have corrected all the lint/flake formatting errors.   Now getting a "check black formatting" which I have been struggling to understand.   Perhaps it is insisting the _MOV and _ACTION arrays to only have 1 value per line?

Anyway, holding off making any more edits until I get Phil's guidance.

GaryOkie committed:
Ran Black formatter on camera.py

all new to me.

GaryOkie committed:
snake bit.  some variables too short.

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -547,3 +600,22 @@ def _start_tour(self, start):
+        try:
+            channel = 1
+            hspeed = 0
+            vspeed = "vertical_speed=1"
```
This will result in `&arg2=vertical_speed=1`. I don't think it's supposed to be that. I think it should just be `&arg2=1`.

GaryOkie replied:
My original code was using python-amcrest's functions of "_move_up(action=None, channel=0, vertical_speed=1_) , etc.   The resulting code was an ugly bunch of nested if's, as I wasn't a good enough Python programmer (after 2 days learning!) to map a _movement_ name to the function name dynamically.   

Then I stumbled on the generic _ptz_control_command_ which was great to dynamically specify actions/codes.   It left me a bit confused how to handle the remaining arguments so I tried to pattern them after how the named "move_xxxx" functions were handling them.   Apparently I got this wrong, even though what I had actually worked!  

Thanks for all the corrections and advice!

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -51,6 +52,43 @@
+    "RightUp",
+    "LeftDown",
+    "LeftUp",
+]
```
To simply the method below, I'd break these into subgroups:
```python
_ZOOM_ACTIONS = ["ZoomWide", "ZoomTele"]
_MOVE_1_ACTIONS = ["Right", "Left", "Up", "Down"]
_MOVE_2_ACTIONS = ["RightDown", "RightUp", "LeftDown", "LeftUp"]
_ACTION = _ZOOM_ACTIONS + _MOVE_1_ACTIONS + _MOVE2_ACTIONS
```

GaryOkie replied:
I originally had the arrays defined on multiple continuation lines, but black formatter split them as above, 1 value per line.    I like what you did here, just hope black leaves it be.

pnbruckner replied:
I didn't make this suggestion to minimize the number of source lines. The reason was to make it easier to determine the arguments for the call to `ptz_control_command`. It's just a happy coincidence that it make the code more compact, too. :wink:

GaryOkie replied:
Got it!   Black left it exactly as is.   I have all the code changes implemented and TESTED now - works great.    Just one _MOVE2_ACTION typo to correct was all.    Updating git now.

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -547,3 +600,22 @@ def _start_tour(self, start):
+            hspeed = 0
+            vspeed = "vertical_speed=1"
+            if "zoom" in movement:
+                vspeed = 0
```
I would do it this way:
```python
if code in _ZOOM_ACTIONS:
    arg1 = arg2 = 0
elif code in _MOVE_1_ACTIONS:
    arg1 = 0
    arg2 = 1
else:
    arg1 = arg2 = 1
```

pnbruckner replied:
FWIW, here's an alternative way to do this:
```python
kwargs = {"code": code, "arg1": 0, "arg2": 0, "arg3": 0}
if code in _MOVE_1_ACTIONS:
    kwargs["arg2"] = 1
elif code in _MOVE_2_ACTIONS:
    kwargs["arg1"] = kwargs["arg2"] = 1
```
Then below:
```python
self._api.ptz_control_command(action="start", **kwargs)
```

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -547,3 +600,22 @@ def _start_tour(self, start):
+            if "zoom" in movement:
+                vspeed = 0
+
+            self._api.ptz_control_command(channel, "start", code, hspeed, vspeed)
```
`channel` should be `0`. Also, I think you need to specify `arg3`, otherwise it will result in `&arg3=None`, when it should be `&arg3=0`.

I would just do it this way:
```python
            self._api.ptz_control_command(action="start", code=code, arg1=arg1, arg2=arg2, arg3=0)
```

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -406,6 +455,10 @@ def disable_motion_detection(self):
         """Call the job and stop camera tour."""
         await self.hass.async_add_executor_job(self._start_tour, False)
 
+    async def async_ptz_control(self, movement, travl_time):
```
Why don't you just use `travel_time`?

pnbruckner reviewed (changes requested):
Very good start. Some fixes/improvements requested.

GaryOkie commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -51,6 +52,43 @@
+    "bottom_right",
+    "top_right",
+    "bottom_left",
+    "top_left",
```
It's really not necessary at all to have this duplicated.   I originally wrote this to use the same terminology in the API (right_up, left_down...) which seems more natural to me.   But then I later discovered that the Foscam component chose to match these directions to the names of their associated mdi icon arrows  (top_right, bottom_left...).

It was simple enough to chose to accept either term in the code.   If you think that is just not needed, then please pick which 4 terms you prefer, and we'll just go with that.

pnbruckner replied:
I tend to think it makes more sense for the values of the `movement` parameter to be consistent. If we're using `left`, `right`, `up` and `down`, then the combined actions should use those terms rather than something different.

GaryOkie replied:
That was my original thinking too and of course that matches the API.      I just now went the other way when examining the lovelace card for handling the movement.     I can easily see someone who uses Foscam (very popular before Amcrest split) getting an Amcrest and appreciate the consistency.    Really no big deal.   I can change it back.

pnbruckner replied:
I kind of hate to ask, but please change it back. Besides the consistency with the other movement options, down & up are directions, whereas bottom and top are locations. And since this a parameter which indicates movement direction...

GaryOkie replied:
ok - changed it back to how I had it in the first place!

GaryOkie commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -51,6 +52,43 @@
+    "top_left",
+]
+_ACTION = [
+    "ZoomTel",
```
nice typo catch!    I don't have an optical zoom camera, so zoom unfortunately wasn't tested.   Do you?

pnbruckner replied:
I honestly don't remember if the PTZ camera I have does optical zoom, but I know it supports zooming (either optically or digitally), so I would guess it would support zooming this way. I'll give this stuff a try after this first review round and you check in the requested changes.

GaryOkie committed:
Implemented Phil's corrections and suggestions.   

Removed duplicated angular movements and code now consistent with Foscam documentation.

GaryOkie committed:
Update angular movements for consistency

Now consistent with MDI Arrow icon names and Foscam component

GaryOkie commented:
Lint/isort didn't like where I inserted  _import homeassistant.helpers.config_validation as cv_  after the other imports?
Not understanding this failed check at all.

pnbruckner commented:
Are you doing the commits via github, or have you installed the dev environment and doing the commit from the command line? If you do it via the second method it generally won't let you check anything in that will fail those tests. Did you read: https://developers.home-assistant.io/docs/development_environment ?

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -4,6 +4,7 @@
 from functools import partial
 import logging
 import time
+import homeassistant.helpers.config_validation as cv
```
This line should go before:
```python
from homeassistant.helpers.dispatcher import async_dispatcher_connect
```

GaryOkie replied:
Yes, thanks - I had just figured that out from the raw log and had already moved the statement.

Yes, I am doing commits directly from github.   And as you can surmise from all the previous failed commits, my journey with trial and error fixing lint, flake, black has been an interesting one.   

I don't have a typical dev environment, but it sure looks like I am needing to go that way if I am ever to improve my progress with component development.    This was just me working from home playing around to see if I could accomplish some useful new feature  that I might interest you in incorporating into a later PR.   I really wasn't prepared to take on a new PR with all the development requirements that entails, but am starting to pick up this stuff as I go, bit by bit.

pnbruckner replied:
Absolutely no problem. I started kind of the same way, but you seem to be progressing way faster than I did. lol! BTW, when I said it was best for you to check the changes into a branch on your own fork, you could have stopped there and just sent me a link to it. But it's good that you did submit the PR. You might as well get credit.

GaryOkie committed:
small tweaks per Phil's comments

re positioned import cv statement.
Replaced logic incorporating kwargs technique for cleaner code.

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -547,3 +586,22 @@ def _start_tour(self, start):
+            if code in _MOVE_1_ACTIONS:
+                kwargs["arg2"] = 1
+            elif code in _MOVE_2_ACTIONS:
+                kwargs["arg1"] = kwargs["arg2"] = 1
```
These statements should be outside (i.e., before) the `try` block. (I just got called out for this exact same thing on my PR. :smile: )

GaryOkie replied:
fixed.   thanks,

pnbruckner commented on homeassistant/components/amcrest/services.yaml:
```diff
@@ -73,3 +73,16 @@ stop_tour:
+      description: "Name of the camera, or 'all' for all cameras."
+      example: 'camera.house_front'
+    movement:
+      description: "up, down, right, left, top_right, bottom_right, top_left, bottom_left, zoom_in, zoom_out"
```
These should change, too.

pnbruckner commented on homeassistant/components/amcrest/services.yaml:
```diff
@@ -73,3 +73,16 @@ stop_tour:
+      description: "up, down, right, left, top_right, bottom_right, top_left, bottom_left, zoom_in, zoom_out"
+      example: 'right'
+    travel_time:
+      description: "(optional) Incremental travel time in fractional seconds: from 0 to 1.  Default:  .2"
```
Why "Incremental"? Not sure what you mean by that. Why not just "Travel time..."?

GaryOkie replied:
agree it is not meaningful.    removed.

pnbruckner reviewed (changes requested):
After these few changes I think we're good. I'd like to hold off merging these changes, though, until the other PR is resolved if that's ok.

GaryOkie committed:
final tweaks (hopefully)

move kwargs stuff above try block.
reverted top/bottom to up/down

pnbruckner commented:
Still need to update `services.yaml` to reflect change back to `right_down`, etc.

GaryOkie committed:
2 blank lines had whitespace

GaryOkie committed:
revert  top/bottom angular directions

pnbruckner commented:
Thanks! It's looking good now. I'll try it out with my PTZ camera. Assuming that works as expected, and once my streaming event PR has been merged, I'll merge this one.

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -51,6 +53,28 @@
+_ATTR_PTZ_MOV = "movement"
+_MOV = [
+    "zoom_in",
+    "zoom_out",
```
These two are backwards. The first (that goes with `ZoomWide`) should be `zoom_out`, and the second (that goes with `ZoomTele`) should be `zoom_in`.

GaryOkie replied:
fixed

pnbruckner commented:
Also, FWIW, I had both the live view in HA showing (in a separate browser from where I was issuing the services) and the camera displayed in the Amcrest app, and the movement was instantaneous on the Amcrest app, but there was a 7 to 8 second delay in HA's live view. I think that's unfortunately just the nature of the `stream` component.

GaryOkie committed:
corrected zoom_in/out order

GaryOkie commented:
Thanks for the testing Phil!    Yes, we'll just have to live with the lag.   My  documentation update says to expect a lag of "several" seconds, and per your test, that is a reasonable statement.   I'm just happy to be able to accurately reposition the camera now with HA without having to use other means.   PTZ presets works fine for many use cases, but not all.

pnbruckner commented:
I think that's everything. Now we just need the reviewers of my PR to approve & ...

GaryOkie commented:
just trying to clear the "Changes requested" when I believe I have made and committed all the changes requested.    Sorry if asking to re-review wasn't appropriate when you said everything was fine, just guessing there is a step I missed to have this cleared.

pnbruckner commented:
No problem. I'm waiting to do that (and add my approval) until my other PR is done. I was trying to avoid someone else coming along and merging this beforehand. Although I'm allowed to approve and merge someone else's PR, I'm not allowed to do that on my own (even though they are both for the same component for which I'm the "code owner.")

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -406,6 +441,10 @@ def disable_motion_detection(self):
 
+    async def async_ptz_control(self, movement, travel_time):
+        """Call the job and move/zoom PTZ camera."""
+        await self.hass.async_add_executor_job(self._ptz_control, movement, travel_time)
```
```suggestion
        """Move or zoom camera in specified direction."""
        code = _ACTION[_MOV.index(movement)]

        kwargs = {"code": code, "arg1": 0, "arg2": 0, "arg3": 0}
        if code in _MOVE_1_ACTIONS:
            kwargs["arg2"] = 1
        elif code in _MOVE_2_ACTIONS:
            kwargs["arg1"] = kwargs["arg2"] = 1

        try:
            await self.hass.async_add_executor_job(
                partial(self._api.ptz_control_command, action="start", **kwargs)
            )
            await asyncio.sleep(travel_time)
            await self.hass.async_add_executor_job(
                partial(self._api.ptz_control_command, action="stop", **kwargs)
            )
        except AmcrestError as error:
            log_update_error(
                _LOGGER, "move", self.name, f"camera PTZ {movement}", error
            )
```

pnbruckner commented on homeassistant/components/amcrest/camera.py:
```diff
@@ -547,3 +586,23 @@ def _start_tour(self, start):
+        except AmcrestError as error:
+            log_update_error(
+                _LOGGER, "move", self.name, f"camera PTZ {movement}", error
+            )
```
```suggestion
```

pnbruckner reviewed (changes requested):
Move non-async code into async method so a worker thread is not tied up during `travel_time` delay.

GaryOkie committed:
Apply suggestions from code review

Co-Authored-By: Phil Bruckner <pnbruckner@gmail.com>

pnbruckner commented:
Sorry, my bad, I forgot to run the checks. Looks like you need to remove `import time`.

GaryOkie committed:
remove now-unused import time

pnbruckner commented:
Thank you! I'll approve now and once the tests pass I'll merge.

GaryOkie commented:
yep, already did.   should fly now.  Thanks!
